tests/firedrake/ensemble/test_ensemble_function.py

```diff
@@ -0,0 +1,153 @@
+    fd.MixedElement([scalar_elements[e] for e in ('T-DG',)]),  # 2
+    scalar_elements['T-DG'],  # 3
+    scalar_elements['BDM'],  # 4
+    scalar_elements['BDM'],  # 5
+    fd.MixedElement([scalar_elements[e] for e in ('BDM', 'V-RT', 'V-RT')])  # 6
+]
+
+# Test with four time partitions:
+# time partitions - serial
+# time partitions - parallel with P<N
+# time partitions - parallel with P=N
+# Also test with spatial parallelism to check that we don't mix up comms accidentally somewhere.
+#
+#       time      space      partition
+# P1 : serial   | serial   | 6
+# P2 : parallel | serial   | 2, 4
+# P4 : parallel | parallel | 3, 3
+# P6 : parallel | serial   | 1 per rank
+
+
+@pytest.fixture
+def Wlocal(ensemble, mesh):
+    gcomm = ensemble.global_comm
+    erank = ensemble.ensemble_rank
+    esize = ensemble.ensemble_size
+    if gcomm.size == 1:
+        elems = elements
+    elif gcomm.size == 2:
+        if erank == 0:
+            elems = elements[:2]
+        if erank == 1:
+            elems = elements[2:]
+    elif gcomm.size == 4:
+        nelem = len(elements)//esize
+        offset = erank*nelem
+        elems = elements[offset:offset+nelem]
+    elif gcomm.size == 6:
+        elems = [elements[erank]]
+    else:
+        raise ValueError("Invalid number of ranks")
+
+    return [fd.FunctionSpace(mesh, e) for e in elems]
+
+
+space_type = ["primal", "dual"]
+
+
+@pytest.fixture(params=space_type)
+def ensemblespace(request, ensemble, Wlocal):
+    if request.param == 'primal':
+        return fd.EnsembleFunctionSpace(Wlocal, ensemble)
+    elif request.param == 'dual':
+        return fd.EnsembleDualSpace([V.dual() for V in Wlocal], ensemble)
+    else:
+        raise ValueError(f"Unknown function space type {request.param}")
+
+
+@pytest.fixture
+def ensemblefunc(ensemblespace):
+    return fd.EnsembleFunction(ensemblespace)
+
+
+@pytest.mark.parallel(nprocs=[1, 2, 4, 6])
+def test_zero(ensemblefunc):
+    """
+    Test setting all components to zero.
+    """
+    # assign some nonzero value
+    assign_scalar(ensemblefunc, 1)
+
+    # check the norm is nonzero
+    for u in ensemblefunc.subfunctions:
+        assert norm(u) > 1e-14, "This test needs a nonzero initial value."
+
+    ensemblefunc.zero()
+
+    for u in ensemblefunc.subfunctions:
+        assert norm(u) < 1e-14, "EnsembleFunction.zero should zero all components"
+
+
+@pytest.mark.parallel(nprocs=[1, 2, 4, 6])
+def test_zero_with_subset(ensemblefunc):
+    """
+    Test setting a subset of all components to zero.
+    """
+    from pyop2 import Subset
+
+    # assign some nonzero value
+    nonzero = 1
+    assign_scalar(ensemblefunc, nonzero)
+
+    subsets = [None if type(V.ufl_element()) is fd.MixedElement else Subset(V.node_set, [0, 1])
+               for V in ensemblefunc.function_space().local_spaces]
+
+    ensemblefunc.zero(subsets)
+
+    for u, subset in zip(ensemblefunc.subfunctions, subsets):
+        if subset is None:
+            continue
```

Comment:
why not assert something for this case?

Reply:
A `Function` in a `MixedFunctionSpace` won't accept a `subset` argument to `assign` or `zero`, so on line 144 I only create a `Subset` for non-mixed components.

I'll add a comment explaining what is going on.